package/MDAnalysis/analysis/helix_analysis.py

```diff
@@ -0,0 +1,474 @@
+# MDAnalysis: A Python package for the rapid analysis of molecular dynamics
+# simulations. In S. Benthall and S. Rostrup editors, Proceedings of the 15th
+# Python in Science Conference, pages 102-109, Austin, TX, 2016. SciPy.
+# doi: 10.25080/majora-629e541a-00e
+#
+# N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein.
+# MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations.
+# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
+#
+
+"""
+HELANAL --- analysis of protein helices
+=======================================
+
+:Author: Lily Wang
+:Year: 2020
+:Copyright: GNU Public License v3
+
+.. versionadded:: 1.0.0
+
+This module contains code to analyse protein helices using the 
+`HELANAL <http://nucleix.mbu.iisc.ernet.in/helanal/helanal/helanal.html>`_ algorithm
+([Bansal2000]_ , [Sugeta1967]_ ).
+
+`HELANAL <http://nucleix.mbu.iisc.ernet.in/helanal/helanal/helanal.html>`_ 
+quantifies the geometry of helices in proteins on the basis of their Cα
+atoms. It can determine local structural features such as the local helical twist and
+rise, virtual torsion angle, local helix origins and bending angles between
+successive local helix axes. 
+
+Example use
+-----------
+
+You can pass in a single selection::
+
+    import MDAnalysis as mda
+    from MDAnalysis.tests.datafiles import PSF, DCD
+    from MDAnalysis.analysis import helix_analysis as hel
+    u = mda.Universe(PSF, DCD)
+    helanal = hel.HELANAL(u, select='name CA and resnum 161-187')
+    helanal.run()
+
+    print(helanal.summary)
+
+Alternatively, you can analyse several helices at once by passing 
+in multiple selection strings::
+
+    helanal2 = hel.HELANAL(u, select=('name CA and resnum 100-160',
+                                      'name CA and resnum 200-230'))
+
+The :func:`helix_analysis` function will carry out helix analysis on 
+atom positions, treating each row of coordinates as an alpha-carbon 
+equivalent::
+
+    hel_xyz = hel.helix_analysis(u.atoms.positions, ref_axis=[0, 0, 1])
+
+"""
+from __future__ import division, absolute_import
+
+import warnings
+import numpy as np
+
+import MDAnalysis as mda
+from ..lib import util, mdamath
+from .base import AnalysisBase
+
+
+def vector_of_best_fit(coordinates):
+    """Fit vector through the centered coordinates, 
+    pointing to the first coordinate (i.e. upside-down).
+
+    Parameters
+    ----------
+    coordinates: :class:`numpy.ndarray` of shape (N, 3)
+
+    Returns
+    -------
+    :class:`numpy.ndarray` of shape (3,)
+        Vector of best fit.
+    """
+    centered = coordinates - coordinates.mean(axis=0)
+    Mt_M = np.matmul(centered.T, centered)
+    u, s, vh = np.linalg.linalg.svd(Mt_M)
+    vector = vh[0]
+
+    # does vector face first local helix origin?
+    angle = mdamath.angle(centered[0], vector)
+    if angle > np.pi/2:
+        vector *= -1
+    return vector
+
+
+def local_screw_angles(global_axis, ref_axis, helix_directions):
```

Comment:
The screw angles here are calculated in the range 0 to pi, but I think they should cover the full -pi to pi range?

Reply:
I guess I thought of it just as the angle to the plane of `ref_axis` and `global_axis`, which is usually [0, pi). I'm really awful at 3D visualisation so I could be missing an obvious answer to this, but which side of the plane is the "origin" i.e. which part of the helix should be negative vs positive pi?

Reply:
So the way I understand it is that the screw angle is measuring the location of each CA around the circle if you were looking down the helix axis - so you could use it to, say, select all the residues on a particular 'face' of the helix by defining an angle range. So positive/negative angles tell us whether it's clockwise or anticlockwise relative to the reference axis.

On the calculation of the screw angles - I don't seem to be getting the numbers I expect, with offsets in multiples of 90/flipped signs. I'm still gathering my thoughts and making sure what I'm expecting is actually right, but for what it's worth the following does seem to work so far (adding back ortho, which I believe is the projection of `ref_axis` on the helix cross-section plane, as the second reference; and changing how the angle is adjusted for quadrants).

    orth = np.cross(-plane, global_axis)  ### 

    # angles to plane normal
    refs = np.array([plane, orth]) ###
    norms = np.outer(mdamath.pnorm(refs), mdamath.pnorm(helix_directions))
    cos = np.matmul(refs, helix_directions.T)/norms
    to_plane, to_orth = np.arccos(np.clip(cos, -1, 1))  # (2, n_vec)

    # angles to plane in (-pi, pi]
    to_orth[to_plane > np.pi/2] *= -1

    return np.rad2deg(to_orth)

Reply:
Thanks @fiona-naughton for looking at it, I remain a bit confused as to what this should be. I added some tests that involve circles as a kind of expected spec: a circle in the xy plane with the x-axis as global_axis and y-axis as ref_axis should have 0s as the angles, and a circle in the xz plane should be `array([   0,   12,   24,   36,   48,   60,   72,   84,   96,  108,  120,
        132,  144,  156,  168,  180, -168, -156, -144, -132, -120, -108,
        -96,  -84,  -72,  -60,  -48,  -36,  -24,  -12])`. I interpreted this function as calculating the angle to the plane of ref_axis and global_axis (i.e. the xy plane)

Reply:
Ah, hadn't looked properly at the new tests, sorry. So having looked through everything again and having another thing, I'm going to gather my thoughts here, just so we can make sure we're on the same page - might get a bit long, so please bear with me!

If I'm following correctly, basically what I think the current code is doing is: **1.** Define an axis (`plane` in the code, but I'll call it `perp` here) perpendicular to both `ref` and `global`. **2.** Measure angles of position vectors to `perp` and `global`.  **3.** Work out quadrants from these angles to give an angle to `global` in the range -pi to pi. 
We get an angle *to* `global `- I guess this would represent how high/low (along the helix axis) an atom is relative to the center of it's helix window? However, using both angles/quadrants like this assumes that the position vectors are ~in the `global`-`perp` plane, which is the case in your second example. I think the first still works because it's a trivial case, elsewhere it'd get less accurate.


I think what it should do is: **1.** Define an axis (`perp`) perpendicular to both `ref` and `global`. **2.** Define a second axis (`orth` perpendicular to both `perp` and `global`). **3.** Measure angles of position vectors to `perp` and `orth`. **4.** Work out quadrants from these angles to give an angle to `orth` in the range -pi to pi.
Since `perp` and `orth` are both perpendicular to `global`, this gives an angle *about* `global` (with 0 in some way based on the direction of `ref`), so represents which side of the helix the atom is on. As above this assumes that the position vectors are in the `perp`-`orth` plane - i.e. perpendicular to `global`. But in general helanal already assumes this is ~the case, so here I think it's (more) ok.


(Related: I also used a circle in xy to test the code, but defined `global` along the z-axis so the position vectors were perpendicular to global as they would be for a 'proper' helix. I don't think it makes sense to test with `global` in the plane of the circle). 

I hope that all makes sense...!

**tl;dr** code currently calculated angle *to* helix axis, but the method for this makes an assumption; tests pass because they're cases where assumption holds. An angle *around* helix axis should require less assumption + more intuitive?

Reply:
Ok, I've just drawn this terribly, just pretend ortho is perpendicular to `perp` and `GLOBAL`:

![IMG_3282 JPG](https://user-images.githubusercontent.com/31115101/84124234-4a74ef00-aa7e-11ea-9bdf-6eb4e8b00566.jpeg)


> Work out quadrants from these angles to give an angle to `orth` in the range -pi to pi.

So it sounds like you think the screw angles should be the blue psi angle to the `ortho` vector, whereas I've been interpreting it as the green theta angle to the plane of REF and GLOBAL. Does that look about right?

> Ah, hadn't looked properly at the new tests, sorry. So having looked through everything again and having another thing, I'm going to gather my thoughts here, just so we can make sure we're on the same page - might get a bit long, so please bear with me!
> 
> If I'm following correctly, basically what I think the current code is doing is: **1.** Define an axis (`plane` in the code, but I'll call it `perp` here) perpendicular to both `ref` and `global`. **2.** Measure angles of position vectors to `perp` and `global`. **3.** Work out quadrants from these angles to give an angle to `global` in the range -pi to pi.
> We get an angle _to_ `global `- I guess this would represent how high/low (along the helix axis) an atom is relative to the center of it's helix window? However, using both angles/quadrants like this assumes that the position vectors are ~in the `global`-`perp` plane, which is the case in your second example. I think the first still works because it's a trivial case, elsewhere it'd get less accurate.

What I *think* I'm doing is figuring out the angle to `perp`, the normal of the ref-global plane, and also figuring out the angle to global as way to get quadrants. The final result after some arithmetic I *think* is the angle to the ref-global plane. I envisioned vectors rotating around the `global` axis being calculated to a static plane, I guess. 

> so we can make sure we're on the same page

I'm happy to be on either page. I don't follow that getting angles to the cross section of the axis (as defined by either `perp` or `ortho`) is the right thing here though, does that capture rotation around `global`?

> (Related: I also used a circle in xy to test the code, but defined global along the z-axis so the position vectors were perpendicular to global as they would be for a 'proper' helix. I don't think it makes sense to test with global in the plane of the circle).

I think I made the circle in the xz-plane, the global axis the x-axis, and the ref-axis the y-axis to test angles to the xy plane. The position vectors in a real helix would rarely be perpendicular because the helix goes up/down as it winds around.

> I hope that all makes sense...!

I think we're making progress? I won't change anything just yet until we settle just what the screw angles represent, though

Reply:
Looking at it again I think the best angle would be from the red vector to the plane of `perp` and `global` (not on the picture)?

Reply:
Ok, I did what was in the last comment and added a lot more comments to the tests to try to make it clear what I'm going for

Reply:
> I think I made the circle in the xz-plane, the global axis the x-axis, and the ref-axis the y-axis to test angles to the xy plane. The position vectors in a real helix would rarely be perpendicular because the helix goes up/down as it winds around.

Using positions here was probably misleading - I mean the `local_helix_directions`, which are what is passed into the `local_screw_angles` calculation. They're calculated as the normalised `bisectors`, which should be roughly perpendicular to the helix axis (e.g. if I feed the test helix along the z axis `np.array([[1.,0,0], [0,1,1], [-1,0,2], [0,-1,3], [1,0,4], [0,1,5], [-1,0,6],[0,-1,7]])` into `helix_analysis`, the `local_helix_directions` are `[[0.  1.  0.],  [-1.  0.  0.], [ 0. -1.  0.], [ 1.  0.  0.], [ 0.  1.  0.], [-1.  0.  0.]]` - ie. in all in the xy plane. 

When trying with an actual helix it's not as perfect, but still close - e.g.

    u = mda.Universe(PDB_small)
    sel = u.select_atoms('name CA and resnum 161-170')
    pos = sel.atoms.positions
    data = hel.helix_analysis(pos)
    #  get angle between global and each helix direction
    print([np.rad2deg(np.arccos(np.dot(i, c['global_axis']))) for i in c['local_helix_directions']])

Gives `[89.38454, 90.11533, 85.983376, 93.158035, 89.046326, 89.97367, 89.75803, 91.40081]`. 

That being said - I suppose it's not unreasonable to test the extreme cases to check the code even if it might not reflect a 'realistic' situation. It might actually come up for quite bendy/kinked helicies, but other parts of helanal currently aren't great for those situations, either.

> So it sounds like you think the screw angles should be the blue psi angle to the ortho vector, whereas I've been interpreting it as the green theta angle to the plane of REF and GLOBAL. Does that look about right?

Yeah, that diagram looks right for what I think we were each considering, though I had been mistakenly assuming you meant the green angle to represent the angle to the `global` axis so my comments about inaccuracy don't matter - my bad there. 

Here's my own diagram - I'm sorry it got a bit busy, especially after adding the new angle-to-perp/global in purple, labeled alpha (though I accidentally drew it going from orth rather than xyz), but hopefully you can still sort of see what's going on.

![IMG_20200610_222617](https://user-images.githubusercontent.com/10507372/84348696-7077bd80-ab6a-11ea-9870-7296d2ec0a1c.jpg)

So I could probably have been clearer that what I think the 'real' screw angle should be is the pink angle, which is basically a projection of the angle between ref and the helix direction vector in the plane perpendicular to global. But I've been talking about the blue angle as an approximation to this, since this was what I interpreted the old code as doing - but we should be able to directly calculate it rather than using the approximate. 

It seems both of what we're doing approximate to the same thing when we assume the helix directions are perpendicular to global (or well, with a quadrant offset, but since where we set 0 is a bit arbitrary this isn't huge) - and that's what I'm seeing testing the current code, so that's something (I was getting the occasional weird thing - like angles above 360 - previously). 

So I guess where I am at the moment is that the code currently calculates approximately the same thing that the original helanal code did/what I was trying to argue, just that it may not be the most 'intuitive' method given what I think the screw angles is trying to represent (though that might just be for me personally); and that I probably should have been advocating for the pink angle all along rather than the blue angle.

Reply:
Thanks so much for drawing it out and clearing things up @fiona-naughton! In my last commit we were approximating to the same thing when you were talking about the blue angle to ortho, but you're right, it makes way more sense to just directly calculate the pink psi angle of the projection onto the ortho-perp plane.

I've just pushed a commit that (I think):

- projects the input vectors onto the global axis
- subtracts the global_axis projection from the input vectors, giving the projection onto the plane
- calculates angles from the projection to `ortho`
- uses the angles to `perp` to assign quadrants

and tests that check that changing the direction of the input vectors, along the global_axis axis, do not result in different angles.

Thank you for your patience in sorting this out with me <3

Reply:
That makes sense, and the it's looking good to me! I might do one last quick lookover but hopefully this is everything :)

(And thank you, too! Now, here's hoping we don't have to think about 3D vector geometry again any time soon... )

Reply:
Thank you @fiona-naughton for all the work you've put into this, I really would have had no idea what to do here without you! I'm definitely looking forward to giving my lack of 3D spatial visualisation skills a break.